### ISSUE TITLE
Benchmark methods separately

### ISSUE THREAD

**Issue by wlandau** [mlr-org/mlr-outreach#2]
In https://github.com/mlr-org/mlr-outreach/tree/master/2019_warsaw,  `benchmark()` is called on full design, so all the methods are run inside a single target. What if we called `benchmark()` separately on each of `svn`, `knn`, etc? Advantages:

1. Parallel efficiency. We get to spread the benchmarks across multiple targets, which lets us use within-target multicore parallelism for something else.
2. We can add new models and modify old ones without invalidating the entire benchmark suite.

All the targets except `bm` are super quick. So essentially, all `drake` can do is decide whether to run everything or run nothing.

**Comment by wlandau** [comment]
My recommendations:

1. Create a separate target for each combination of learner and task. Reasons:
    1. It distributes the runtime more evenly across jobs.
    2. We can change one of the classifiers and without invalidating the rest. For example, we can change the `kknn()` function, and the `bm_*_svm` targets will stay up to date. And we can add classifiers with impunity.
2. Let's try not to use targets for quick tasks like creating designs.
3. Put code into custom functions.
4. Use `drake_plan()` instead of `code_to_plan()`. Things become much more concise if we use the [DSL](https://ropenscilabs.github.io/drake-manual/plans.html#large-plans).

I have included a sketch below. Issues:

1. How do we suppress the "INFO [...]" messages? The `show.info` argument was removed in `mlr3`, and I am not sure what to supply to `mlr_control()`.
2. How do we combine multiple `BenchmarkResult` objects into a single `BenchmarkResult` object? I think the `agg_*` targets should be `BechmarkResult`s, not lists.

``` r
suppressPackageStartupMessages({
  library(drake)
  library(future.callr)
  library(mlr3)
  library(mlr3learners)
  library(mlr3tuning)
  library(paradox)
  library(R.utils)
})

kknn <- function(task, measures) {
  task <- mlr_tasks$mget(task)
  learner <- mlr_learners$mget("classif.kknn")
  measures <- mlr_measures$mget(measures)
  param_set <- ParamSet$new(params = list(
    ParamInt$new("k", lower = 10, upper = 50),
    ParamDbl$new("distance", lower = 1, upper = 50)
  ))
  terminator <- TerminatorEvaluations$new(1) # Increase to scale up.
  resampling_inner <- mlr_resamplings$get("cv3")
  at <- AutoTuner$new(
    learner$classif.kknn,
    resampling_inner,
    measures = measures,
    param_set,
    terminator,
    tuner = TunerRandomSearch
  )
  at$store_bmr <- TRUE
  resampling_outer <- mlr_resamplings$get(
    "repeated_cv",
    param_vals = list(folds = 2, repeats = 1) # Increase to scale up.
  )
  design <- expand_grid(task, list(at), resampling_outer)
  benchmark(design)
}
  
svm <- function(task, measures) {
  task <- mlr_tasks$mget(task)
  measures <- mlr_measures$mget(measures)
  learner <- mlr_learners$mget("classif.svm")
  param_set <- ParamSet$new(params = list(
    ParamDbl$new("cost", lower = -5, upper = 5),
    ParamDbl$new("gamma", lower = -5, upper = 3)
  ))
  param_set$trafo <- function(x, param_set) {
    x$cost = 2^x$cost
    x$gamma = 2^x$gamma
    return(x)
  }
  terminator <- TerminatorEvaluations$new(1) # Increase to scale up.
  resampling_inner <- mlr_resamplings$get("cv3")
  at <- AutoTuner$new(
    learner$classif.svm,
    resampling_inner,
    measures = measures,
    param_set,
    terminator,
    tuner = TunerRandomSearch
  )
  at$store_bmr <- TRUE
  resampling_outer <- mlr_resamplings$get(
    "repeated_cv",
    param_vals = list(folds = 2, repeats = 1) # Increase to scale up.
  )
  design <- expand_grid(task, list(at), resampling_outer)
  benchmark(design)
}

# But actually benchmarking classifiers is harder,
# so we put them in the plan
plan <- drake_plan(
  # Run the benchmarks for a single learner and task.
  bm = target(
    learner(task = task, measures = "classif.ce"),
    transform = cross(task = c("iris", "spam"), learner = c(kknn, svm))
  ),
  # Aggregate the results by task.
  agg = target(
    list(bm), # To do: return a BenchmarkResult, not a list.
    transform = combine(bm, .by = task) # Remove .by to combine everything.
  )
)

plan
#> # A tibble: 6 x 2
#>   target       command                                     
#>   <chr>        <expr>                                      
#> 1 bm_iris_kknn kknn(task = "iris", measures = "classif.ce")
#> 2 bm_spam_kknn kknn(task = "spam", measures = "classif.ce")
#> 3 bm_iris_svm  svm(task = "iris", measures = "classif.ce") 
#> 4 bm_spam_svm  svm(task = "spam", measures = "classif.ce") 
#> 5 agg_iris     list(bm_iris_kknn, bm_iris_svm)             
#> 6 agg_spam     list(bm_spam_kknn, bm_spam_svm)

drake_plan_source(plan)
#> drake_plan(
#>   bm_iris_kknn = kknn(task = "iris", measures = "classif.ce"),
#>   bm_spam_kknn = kknn(task = "spam", measures = "classif.ce"),
#>   bm_iris_svm = svm(task = "iris", measures = "classif.ce"),
#>   bm_spam_svm = svm(task = "spam", measures = "classif.ce"),
#>   agg_iris = list(bm_iris_kknn, bm_iris_svm),
#>   agg_spam = list(bm_spam_kknn, bm_spam_svm)
#> )

config <- drake_config(plan)
vis_drake_graph(config)
```

![](https://i.imgur.com/3Hg9DYy.png)

``` r

make(plan)
#> target bm_iris_kknn
#> INFO  [14:07:52.821] Benchmark with 2 resampling iterations 
#> INFO  [14:07:52.838] Applying learner 'autotuner' on task 'iris' (iter 1/2) 
#> INFO  [14:07:53.732] Evaluating 1 configurations 
#> INFO  [14:07:53.800] Benchmark with 3 resampling iterations 
#> INFO  [14:07:53.802] Applying learner 'classif.kknn' on task 'iris' (iter 1/3) 
#> INFO  [14:07:53.827] Applying learner 'classif.kknn' on task 'iris' (iter 2/3) 
#> INFO  [14:07:53.845] Applying learner 'classif.kknn' on task 'iris' (iter 3/3) 
#> INFO  [14:07:53.885] Finished benchmark 
#> INFO  [14:07:53.890] Evaluation finished. Remaining: 0 evaluations. 
#> INFO  [14:07:53.918] Applying learner 'autotuner' on task 'iris' (iter 2/2) 
#> INFO  [14:07:54.071] Evaluating 1 configurations 
#> INFO  [14:07:54.130] Benchmark with 3 resampling iterations 
#> INFO  [14:07:54.132] Applying learner 'classif.kknn' on task 'iris' (iter 1/3) 
#> INFO  [14:07:54.151] Applying learner 'classif.kknn' on task 'iris' (iter 2/3) 
#> INFO  [14:07:54.169] Applying learner 'classif.kknn' on task 'iris' (iter 3/3) 
#> INFO  [14:07:54.209] Finished benchmark 
#> INFO  [14:07:54.211] Evaluation finished. Remaining: 0 evaluations. 
#> INFO  [14:07:54.247] Finished benchmark
#> target bm_iris_svm
#> INFO  [14:07:54.544] Benchmark with 2 resampling iterations 
#> INFO  [14:07:54.546] Applying learner 'autotuner' on task 'iris' (iter 1/2) 
#> INFO  [14:07:54.759] Evaluating 1 configurations 
#> INFO  [14:07:54.877] Benchmark with 3 resampling iterations 
#> INFO  [14:07:54.883] Applying learner 'classif.svm' on task 'iris' (iter 1/3) 
#> INFO  [14:07:54.900] Applying learner 'classif.svm' on task 'iris' (iter 2/3) 
#> INFO  [14:07:54.915] Applying learner 'classif.svm' on task 'iris' (iter 3/3) 
#> INFO  [14:07:54.952] Finished benchmark 
#> INFO  [14:07:54.955] Evaluation finished. Remaining: 0 evaluations. 
#> INFO  [14:07:54.971] Applying learner 'autotuner' on task 'iris' (iter 2/2) 
#> INFO  [14:07:55.177] Evaluating 1 configurations 
#> INFO  [14:07:55.293] Benchmark with 3 resampling iterations 
#> INFO  [14:07:55.295] Applying learner 'classif.svm' on task 'iris' (iter 1/3) 
#> INFO  [14:07:55.313] Applying learner 'classif.svm' on task 'iris' (iter 2/3) 
#> INFO  [14:07:55.328] Applying learner 'classif.svm' on task 'iris' (iter 3/3) 
#> INFO  [14:07:55.367] Finished benchmark 
#> INFO  [14:07:55.369] Evaluation finished. Remaining: 0 evaluations. 
#> INFO  [14:07:55.398] Finished benchmark
#> target bm_spam_kknn
#> INFO  [14:07:55.679] Benchmark with 2 resampling iterations 
#> INFO  [14:07:55.681] Applying learner 'autotuner' on task 'spam' (iter 1/2) 
#> INFO  [14:07:55.831] Evaluating 1 configurations 
#> INFO  [14:07:55.893] Benchmark with 3 resampling iterations 
#> INFO  [14:07:55.894] Applying learner 'classif.kknn' on task 'spam' (iter 1/3) 
#> INFO  [14:07:56.349] Applying learner 'classif.kknn' on task 'spam' (iter 2/3) 
#> INFO  [14:07:56.798] Applying learner 'classif.kknn' on task 'spam' (iter 3/3) 
#> INFO  [14:07:57.269] Finished benchmark 
#> INFO  [14:07:57.271] Evaluation finished. Remaining: 0 evaluations. 
#> INFO  [14:07:59.055] Applying learner 'autotuner' on task 'spam' (iter 2/2) 
#> INFO  [14:07:59.190] Evaluating 1 configurations 
#> INFO  [14:07:59.253] Benchmark with 3 resampling iterations 
#> INFO  [14:07:59.255] Applying learner 'classif.kknn' on task 'spam' (iter 1/3) 
#> INFO  [14:07:59.832] Applying learner 'classif.kknn' on task 'spam' (iter 2/3) 
#> INFO  [14:08:00.386] Applying learner 'classif.kknn' on task 'spam' (iter 3/3) 
#> INFO  [14:08:00.950] Finished benchmark 
#> INFO  [14:08:00.952] Evaluation finished. Remaining: 0 evaluations. 
#> INFO  [14:08:03.112] Finished benchmark
#> target bm_spam_svm
#> INFO  [14:08:03.531] Benchmark with 2 resampling iterations 
#> INFO  [14:08:03.533] Applying learner 'autotuner' on task 'spam' (iter 1/2) 
#> INFO  [14:08:03.744] Evaluating 1 configurations 
#> INFO  [14:08:03.860] Benchmark with 3 resampling iterations 
#> INFO  [14:08:03.861] Applying learner 'classif.svm' on task 'spam' (iter 1/3) 
#> INFO  [14:08:04.477] Applying learner 'classif.svm' on task 'spam' (iter 2/3) 
#> INFO  [14:08:05.084] Applying learner 'classif.svm' on task 'spam' (iter 3/3) 
#> INFO  [14:08:05.711] Finished benchmark 
#> INFO  [14:08:05.714] Evaluation finished. Remaining: 0 evaluations. 
#> INFO  [14:08:07.359] Applying learner 'autotuner' on task 'spam' (iter 2/2) 
#> INFO  [14:08:07.598] Evaluating 1 configurations 
#> INFO  [14:08:07.718] Benchmark with 3 resampling iterations 
#> INFO  [14:08:07.720] Applying learner 'classif.svm' on task 'spam' (iter 1/3) 
#> INFO  [14:08:08.397] Applying learner 'classif.svm' on task 'spam' (iter 2/3) 
#> INFO  [14:08:09.082] Applying learner 'classif.svm' on task 'spam' (iter 3/3) 
#> INFO  [14:08:09.795] Finished benchmark 
#> INFO  [14:08:09.798] Evaluation finished. Remaining: 0 evaluations. 
#> INFO  [14:08:11.380] Finished benchmark
#> target agg_iris
#> target agg_spam

build_times()
#> # A tibble: 6 x 4
#>   target       elapsed    user       system    
#>   <chr>        <Duration> <Duration> <Duration>
#> 1 agg_iris     0.243s     0.228s     0.007s    
#> 2 agg_spam     0.46s      0.428s     0.024s    
#> 3 bm_iris_kknn 1.914s     7.579s     0.133s    
#> 4 bm_iris_svm  1.18s      6.918s     0.061s    
#> 5 bm_spam_kknn 7.758s     20.867s    0.216s    
#> 6 bm_spam_svm  8.307s     26.351s    0.261s

config <- drake_config(plan)
vis_drake_graph(config)
```

![](https://i.imgur.com/scNKROq.png)

<sup>Created on 2019-07-31 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by pat-s** [comment]
>    -  Parallel efficiency. We get to spread the benchmarks across multiple targets, which lets us use within-target multicore parallelism for something else.
>    -  We can add new models and modify old ones without invalidating the entire benchmark suite.


These are the exact arguments I always tell people why _drake_ and _mlr_ form a great team together. So yes, that is the way to go. Here is an example from my [research project](https://github.com/pat-s/2019-feature-selection/blob/master/code/06-benchmark-matrix.R). I usually do it that way.

I should have made it more clear earlier that this is the way I teach people how to do it - as mentioned in the drake issue, I just got stuck on the first try doing it the "standard non-drake way" (which should nevertheless also work).

## Questions

> How do we suppress the "INFO [...]" messages? The show.info argument was removed in mlr3, and I am not sure what to supply to mlr_control().

This is not yet implemented :)

> How do we combine multiple BenchmarkResult objects into a single BenchmarkResult object? I think the agg_* targets should be BechmarkResults, not lists.

`BenchmarkResult$combine()` -> see `?BenchmarkResult`

```r
  agg = target(
    list(bm), # To do: return a BenchmarkResult, not a list.
    transform = combine(bm, .by = task) # Remove .by to combine everything.
  )
```

This should be done with `BenchmarkResult$combine()` by aggregating all single results into one big container. 
So we would need to choose an arbitrary main container (maybe we can do better than choosing arbitrary?) and merge all the others into it. Here is a sketch, correct me if this does not work :)

((`mlr3::BenchmarkResult$combine()` is equivalent to `?mlr::mergeBenchmarkResults()` which was a [manual task](https://github.com/pat-s/2019-feature-selection/blob/master/code/061-aggregate.R) for me up until now.))

```r
  agg = target(
    bm_spam_svm$combine(bmr), .
    transform = map(<all existing BMRs>)
  )
```

## Misc

> 
>  - Create a separate target for each combination of learner and task. Reasons:
>      -  It distributes the runtime more evenly across jobs.
>      -  We can change one of the classifiers and without invalidating the rest. For example, we can change the kknn() function, and the bm_*_svm targets will stay up to date. And we can add classifiers with impunity.

Fully agree. However, I would argue that we should exclude most parts (speaking of ParamSets, CV settings, etc) and make them function arguments so that we can put them together like building blocks (taking whatever we want to to create a run of this model). I think model functions like `svm()` from your example should have 

- task
- measures
- CV settings (both inner and outer)
- ParamSets

as arguments. These parts should be defined somewhere else and be supplied via arguments.
This is also how I do it in practice:
1. [Here](https://github.com/pat-s/2019-feature-selection/tree/master/code/05-modeling/paper) I define all the inputs for the models
1. Which are [wrapped](https://github.com/pat-s/2019-feature-selection/blob/d7df860373e973b0f8489ad6c5cfe24bc21aa1ed/code/05-modeling/paper/filter-wrapper.R) into their respective wrappers for clarity (these intermediate wrapping steps could actually go into one function also)
1. [Create the whole BM matrix](https://github.com/pat-s/2019-feature-selection/blob/d7df860373e973b0f8489ad6c5cfe24bc21aa1ed/code/06-benchmark-matrix.R) using `drake::cross()` 

> Let's try not to use targets for quick tasks like creating designs.

Yes, small ones can go into the custom "model function", i.e. `svm()` or `knn()`

> Put code into custom functions.

> Use drake_plan() instead of code_to_plan(). Things become much more concise if we use the DSL.

Maybe you can shine a bit more light onto this one for me.
Until now I am very happy with `code_to_plan()` as I can gather all the R-scripts in `_drake.R` and do not have to wrap everything into `drake_plan()` right from the start.
This just feels cleaner for me and I can tell people that they can continue writing their "plan R scripts" first.
Is there an advantage of the `drake_plan()` approach which I am missing here?

**Comment by mllg** [comment]
> > How do we suppress the "INFO [...]" messages? The show.info argument was removed in mlr3, and I am not sure what to supply to mlr_control().
> 
> This is not yet implemented :)

We use the `lgr` package. You can disable it by getting the mlr3 logger from the package and then set a different threshold:
```r
lgr::get_logger("mlr3")$set_threshold("warn")
```

**Comment by berndbischl** [comment]
> We use the `lgr` package. You can disable it by getting the mlr3 logger from the package and then set a different threshold:

you should rather link him to the mlr3book. we need a short section on the logger there?

**Comment by wlandau** [comment]
Thanks, everyone! `lgr::get_logger("mlr3")$set_threshold("warn")` worked for me.

@pat-s, thanks for pointing me to `BenchmarkResult$combine()`. How about this?

```r
combine_benchmarks <- function(...) {
  args <- list(...)
  lapply(args[-1], args[[1]]$combine)
  args[[1]]
}

plan <- drake_plan(
  bm = target(
    learner(task = task, measures = "classif.ce"),
    transform = cross(task = c("iris", "spam"), learner = c(kknn, svm))
  ),
  agg = target(
    combine_benchmarks(bm),
    transform = combine(bm, .by = task)
  )
)
```

**Comment by wlandau** [comment]
> Fully agree. However, I would argue that we should exclude most parts (speaking of ParamSets, CV settings, etc) and make them function arguments so that we can put them together like building blocks (taking whatever we want to to create a run of this model). I think model functions like svm() from your example should have...

Sure, real situations do require this flexibility. Sounds fine to me.

> Until now I am very happy with code_to_plan() as I can gather all the R-scripts in _drake.R and do not have to wrap everything into drake_plan() right from the start.
This just feels cleaner for me and I can tell people that they can continue writing their "plan R scripts" first.
Is there an advantage of the drake_plan() approach which I am missing here?

With `code_to_plan()`, we lose transformations like `map()`, `cross()` and `combine()`. We could still decompose designs into separate benchmarking runs, but every target needs to be written out by hand, which would be tedious and cumbersome for https://github.com/pat-s/2019-feature-selection/blob/d7df860373e973b0f8489ad6c5cfe24bc21aa1ed/code/06-benchmark-matrix.R. In this situation, a call to `drake_plan()` ends up being much more concise than a script for `code_to_plan()`.

In addition, in `code_to_plan()`, things become targets that really shouldn't be: for example, intermediate copies of `ParamSet`s like in https://github.com/ropensci/drake/issues/961. `ParamSet`s themselves are light, so this is not so bad. But for workflows with gigantic datasets, each target would create a new deep copy of a dataset, and storage would explode.

More broadly speaking, sometimes it is not obvious how to group work into targets. `code_to_plan()` makes this decision for you automatically, and it is not always the right one.

**Comment by berndbischl** [comment]
@pat-s whatever you finally do here, shouldnt this nicely be document somewhere? in the mlr book after warsaw? in a section on parallelization?

**Comment by pat-s** [comment]
@wlandau 	This is a very neat and concise summary why `drake_plan()` > `code_to_plan()`. If this is not already in the docs (I assume it is) it should go there!

The `combine_benchmark()` fun looks a bit hacky on the first look and might not be understood by the average user. But if we can reduce all the effort to these few lines I am very happy with it (including explaining what happens here). 

@berndbischl 	
Definitely. Imo this should go into its own section then, including parts of the workshop and a general guide how to use drake+mlr3 to support reproducible research. However this will take more than one day... 
But yes, this is a great way to rise (for both packages) imo since _drake_ is perfectly suited for ML tasks due to its HPC support.

@wlandau 	
I'll try to come back with a full example after I've resolved my local issues. If you have time/motivation to contribute further, you are of course highly welcome to do so (without any expectations from our side ofc) :duck: 
_drake_ is amazing, thanks a ton.

**Comment by pat-s** [comment]
@wlandau 	
I think I've found an elegant solution. 
By using `map()` instead of `cross()` we can integrate varying ParamSets in a flexible way.
This is not possible with `cross()` since this one creates a grid of all possible combinations - which would result in learner/ParamSet combinations which are not valid.

For BM without varying ParamSets, `cross()` is fine since all other options apply to any kind of learner.
However, I am a bit afraid of the implicit definition of (in this case) `learner()` when using `cross()` - [example](https://github.com/mlr-org/mlr-outreach/blob/0cb38fc7c5cee17cd28968b32fc4aba14971b3f2/2019_warsaw/benchmark_plan.R#L38-L39). 
I think this might confuse users even though its nifty. 

### `map()` approach

```r
  bm = target(

    create_learner(learner = learner, learner_id = learner_id, task = task,
      measures = measures, param_set = param_set, resampling_inner = resampling_inner,
      resampling_outer = resampling_outer, terminator = terminator),

    transform = map(
      learner = c("classif.kknn", "classif.svm", "classif.svm", "classif.kknn"),
      learner_id = c("KKNN-ps1", "SVM-ps1", "SVM-ps2", "KKNN-ps2"),
      param_set = c(param_set_knn1, param_set_svm1, param_set_svm2, param_set_knn2), # from R/param-sets.R
      task = c("iris", "spam"),
      measures = "classif.ce",
      resampling_inner = resampling_inner, # from R/resampling.R
      resampling_outer = resampling_outer, # from R/resampling.R
      terminator = terminator, # from R/tuning.R

      .id = c(learner, task)) # this controls the target name generation
  ),
```

See https://github.com/mlr-org/mlr-outreach/blob/0cb38fc7c5cee17cd28968b32fc4aba14971b3f2/2019_warsaw/benchmark_plan.R for the complete R script.

**Comment by pat-s** [comment]
I am thinking of creating an example repo for mlr3+drake that holds several examples:

- small and easy (train+test without tuning) 
- medium (nested resampling with multiple learners)
- advanced (nested resampl + preproc incl. _mlr3pipelines_)

These examples could then be used as a skeleton by users to have a jump start with mlr3+drake.

**Comment by wlandau** [comment]
> By using map() instead of cross() we can integrate varying ParamSets in a flexible way.
This is not possible with cross() since this one creates a grid of all possible combinations - which would result in learner/ParamSet combinations which are not valid.

I like where this is going. By the way, you can create your own custom `map()` grid with `.data`. The sketch below assumes you don't mind creating a manual `tibble` of param set designations, but we could think about automating this if we can get the name of the learner back from each param set.

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(drake)
library(rlang)
library(tidyr)
#> 
#> Attaching package: 'tidyr'
#> The following objects are masked from 'package:drake':
#> 
#>     expand, gather

grid <- tribble(
  ~learner, ~param_set,
  "svm", "ps_svm_1",
  "svm", "ps_svm_2",
  "kknn", "ps_kknn_1",
) %>%
  mutate(
    param_set_sym = syms(param_set),
    learner_id = paste0(learner, "_", param_set)
  ) %>%
  crossing(task = c("iris", "spam"))

print(grid)
#> # A tibble: 6 x 5
#>   learner param_set param_set_sym learner_id     task 
#>   <chr>   <chr>     <list>        <chr>          <chr>
#> 1 kknn    ps_kknn_1 <sym>         kknn_ps_kknn_1 iris 
#> 2 kknn    ps_kknn_1 <sym>         kknn_ps_kknn_1 spam 
#> 3 svm     ps_svm_1  <sym>         svm_ps_svm_1   iris 
#> 4 svm     ps_svm_1  <sym>         svm_ps_svm_1   spam 
#> 5 svm     ps_svm_2  <sym>         svm_ps_svm_2   iris 
#> 6 svm     ps_svm_2  <sym>         svm_ps_svm_2   spam

plan <- drake_plan(
  bm = target(
    create_learner(
      learner = learner,
      learner_id = learner_id,
      task = task,
      measures = "classif.ce",
      param_set = param_set_sym,
      resampling_inner = resampling_inner,
      resampling_outer = resampling_outer,
      terminator = terminator
    ),
    # The !! is important:
    transform = map(.data = !!grid, .id = c(learner, task)) 
  )
)

print(plan)
#> # A tibble: 6 x 2
#>   target       command                                                     
#>   <chr>        <expr>                                                      
#> 1 bm_kknn_iris create_learner(learner = "kknn", learner_id = "kknn_ps_kknn…
#> 2 bm_kknn_spam create_learner(learner = "kknn", learner_id = "kknn_ps_kknn…
#> 3 bm_svm_iris  create_learner(learner = "svm", learner_id = "svm_ps_svm_1"…
#> 4 bm_svm_spam  create_learner(learner = "svm", learner_id = "svm_ps_svm_1"…
#> 5 bm_svm_iris… create_learner(learner = "svm", learner_id = "svm_ps_svm_2"…
#> 6 bm_svm_spam… create_learner(learner = "svm", learner_id = "svm_ps_svm_2"…

drake_plan_source(plan)
#> drake_plan(
#>   bm_kknn_iris = create_learner(
#>     learner = "kknn", learner_id = "kknn_ps_kknn_1",
#>     task = "iris", measures = "classif.ce", param_set = ps_kknn_1,
#>     resampling_inner = resampling_inner, resampling_outer = resampling_outer,
#>     terminator = terminator
#>   ),
#>   bm_kknn_spam = create_learner(
#>     learner = "kknn", learner_id = "kknn_ps_kknn_1",
#>     task = "spam", measures = "classif.ce", param_set = ps_kknn_1,
#>     resampling_inner = resampling_inner, resampling_outer = resampling_outer,
#>     terminator = terminator
#>   ),
#>   bm_svm_iris = create_learner(
#>     learner = "svm", learner_id = "svm_ps_svm_1",
#>     task = "iris", measures = "classif.ce", param_set = ps_svm_1,
#>     resampling_inner = resampling_inner, resampling_outer = resampling_outer,
#>     terminator = terminator
#>   ),
#>   bm_svm_spam = create_learner(
#>     learner = "svm", learner_id = "svm_ps_svm_1",
#>     task = "spam", measures = "classif.ce", param_set = ps_svm_1,
#>     resampling_inner = resampling_inner, resampling_outer = resampling_outer,
#>     terminator = terminator
#>   ),
#>   bm_svm_iris_2 = create_learner(
#>     learner = "svm", learner_id = "svm_ps_svm_2",
#>     task = "iris", measures = "classif.ce", param_set = ps_svm_2,
#>     resampling_inner = resampling_inner, resampling_outer = resampling_outer,
#>     terminator = terminator
#>   ),
#>   bm_svm_spam_2 = create_learner(
#>     learner = "svm", learner_id = "svm_ps_svm_2",
#>     task = "spam", measures = "classif.ce", param_set = ps_svm_2,
#>     resampling_inner = resampling_inner, resampling_outer = resampling_outer,
#>     terminator = terminator
#>   )
#> )
```

<sup>Created on 2019-08-02 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by wlandau** [comment]
> However, I am a bit afraid of the implicit definition of (in this case) learner() when using cross()

I guess I'm more comfortable with it since I wrote `drake`'s DSL. The DSL works by symbol substitution, so it does not matter whether you are mapping over the name of a function or the value of an argument. This may be difficult for some users to understand, but I find it handy in practice.

``` r
library(drake)
drake_plan(y = target(f(x), transform = map(f = c(g, h))))
#> # A tibble: 2 x 2
#>   target command
#>   <chr>  <expr> 
#> 1 y_g    g(x)   
#> 2 y_h    h(x)
```

<sup>Created on 2019-08-02 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by pat-s** [comment]
> I like where this is going. By the way, you can create your own custom map() grid with .data. The sketch below assumes you don't mind creating a manual tibble of param set designations, but we could think about automating this if we can get the name of the learner back from each param set.

I thought quite some time about whether this becomes too technical for the avg user. I am still not sure if we should do this.
This is nice and neat but I am afraid loosing people at this stage. 

I adapted the skeleton to use `tribble()` and automated the generation of the `learner_id`. In your example they were not unique and this is important for merging them later on. 

Since the string extraction need to be called rowwise, we need `dplyr::rowwise()` which does make the code more readable.
Also I should probably put the learner id creation part into a function to declutter.

```r
  settings_grid <- tribble(
    # write learner/ParamSet grid
    ~learner, ~param_set,
    "classif.svm", "param_set_svm1",
    "classif.svm", "param_set_svm2",
    "classif.kknn", "param_set_knn1",
    "classif.kknn", "param_set_knn2"
  ) %>%
    crossing(task = c("iris", "spam")) %>%
    rowwise() %>% # string extraction won't work otherwise
    mutate(param_set_sym = syms(param_set)) %>%
    # automate "pretty" learner id creation
    mutate(learner_id = paste0(
      toupper(str_split_fixed(learner, "[.]", n = 2)[2]),
      "-",
      paste0("PS", str_extract(param_set, "(\\d)+")),
      "-",
      task)
    ),

# A tibble: 8 x 5
  learner      param_set      task  param_set_sym learner_id   
  <chr>        <chr>          <chr> <list>        <chr>        
1 classif.svm  param_set_svm1 iris  <sym>         SVM-PS1-iris 
2 classif.svm  param_set_svm1 spam  <sym>         SVM-PS1-spam 
3 classif.svm  param_set_svm2 iris  <sym>         SVM-PS2-iris 
4 classif.svm  param_set_svm2 spam  <sym>         SVM-PS2-spam 
5 classif.kknn param_set_knn1 iris  <sym>         KKNN-PS1-iris
6 classif.kknn param_set_knn1 spam  <sym>         KKNN-PS1-spam
7 classif.kknn param_set_knn2 iris  <sym>         KKNN-PS2-iris
8 classif.kknn param_set_knn2 spam  <sym>         KKNN-PS2-spam
```

I guess we're nearly there.
This will become example 2/§ (medium complexity). 
In the first example I would like to not include too much automatic creation of targets but focus on a simple linear and clean `drake_plan()` approach. 

Amazing what you build there and how flexible _drake_ is. 

What bothers me a bit still is the readability of `combine_benchmarks()`. I haven't yet taken a shot but maybe we can make the code a bit more readable? This looks very hacky in the first place and the user might not know where `$combine` comes from.

```r
combine_benchmarks = function(...) {
  args = list(...)
  lapply(args[-1], args[[1]]$combine)
  args[[1]]
}
```

**Comment by pat-s** [comment]
FYI: If you take a look at the repo again, do not get confused. I am giving the workshop with @jakob-r 	and he wrote his own version of the drake pipeline as well. I tried to distinguish both by names. He hasn't yet had any practical exp with _drake_ but exactly this is very helpful in the process because there is 2nd/3rd perspective. We will discuss this further in the next days/weeks.

**Comment by wlandau** [comment]
> What bothers me a bit still is the readability of combine_benchmarks(). I haven't yet taken a shot but maybe we can make the code a bit more readable? This looks very hacky in the first place and the user might not know where $combine comes from.

I agree that it is hacky, especially with the ellipsis. Is there a way around `$combine()` in `mlr3`?

**Comment by berndbischl** [comment]
> I agree that it is hacky, especially with the ellipsis. Is there a way around `$combine()` in `mlr3`?

what exactly do you wanna do? in general, if you have concrete usability questions about mlr3, i guess you should ask them in another "channel" (if @pat-s cannot answer them directly)
i just saw that question but for us its really hard to follow this lengty thread in general.

either ask in our tracker or in our slack?

**Comment by pat-s** [comment]
@berndbischl 	Will is kindly helping me to set up the workshop and we are just trying to find a a nice and user-friendly skeleton for drake and mlr3 which will be easy and welcoming for users to use and understand.

Will is not asking questions because he wants to get into mlr3 but to help me here trying to find the best approach - without any added value for him. I guess the discussion we're having here fits into this repo and should not be discussed in mlr3. Sure its lengthy but there is no way around it :) Doing it here is also more transparent than in a Slack channel (for example). (and yes I know you did not mean slack when you were saying "channel" :) )

**Comment by berndbischl** [comment]
ah well, all ok. i had the impression you were looking for "the best way to do X in mlr3"?

i just wanted so: we are HAPPY to help you guys. as i assume that you - @pat-s - maybe dont know every possible funtion in mlr3. just wanted to say: simply ask then, just maybe not here.

but OTOH i guess you know how our slack works :)

**Comment by pat-s** [comment]
@wlandau 	
The workshop source now lives in https://github.com/mlr-org/mlr3-learndrake.
Happy to receive any suggestions/improvements :tada: 

I'll close here, we can discuss things in the new repo.

**Comment by wlandau** [comment]
Awesome! I will check it out.
